### DISCUSSION TITLE
MPI workers seem to be running on same core

### DISCUSSION THREAD

**Discussion by pmenczel** [mpi4py/mpi4py#458] [Q&A] [answered]
**Summary**
I am trying to parallelize a problem on an HPC using the `mpi4py.futures.MPIPoolExecutor`. In short, the problem I am encountering is that the more tasks I run on a single node, the slower they become. It thus seems like they are all running on the same CPU core.
I do not know whether the problem comes from mpi4py, from the HPC setup, from MPICH or from my own code, and I don't even know how to figure that out. I am hoping that someone here could give me a pointer.

**Details**
The HPC runs a SLURM workload manager and I am using the MPICH module.
For testing purposes, I have written a simple python script that creates either an MPIPoolExecutor (`mpitest.py`) or a ProcessPoolExecutor (`nompi.py`) and starts a number of identical tasks. Each task calculates Fibonacci numbers, and returns how many Fibonacci numbers it was able to generate in 5 seconds. I will append the code below. I get the following results:

* **1 worker** (either script)
Calculates approx. 730k Fibonacci numbers
* **50 workers** (`nompi.py`)
Each worker calculates approx. 100k Fibonacci numbers
* **50 workers** (`nompi.py` with `--cpus-per-task=51`)
Each worker calculates approx. 730k Fibonacci numbers
* **50 workers** (`mpitest.py` with `--ntasks=51` and `--cpus-per-task=1`)
Each worker calculates approx. 100k Fibonacci numbers
Note: the final command of the job script is
```mpirun -np $SLURM_NTASKS python -m mpi4py.futures mpitest.py 50```
I will append the full job scripts below.
* **Multiple nodes** with `mpitest.py`
Result depends on how many workers there are per node

As you can see, the version with MPIPoolExecutor behaves like the no-MPI version does if `cpus-per-task` is not specified.

**More details**
The output of `mpiexec --version` is
```
HYDRA build details:
    Version:                                 3.4.2
    Release Date:                            Wed May 26 15:51:40 CDT 2021
    CC:                              <SNIP>
    Process Manager:                         pmi
    Launchers available:                     ssh rsh fork slurm ll lsf sge manual persist
    Topology libraries available:            hwloc
    Resource management kernels available:   user slurm ll lsf sge pbs cobalt
    Demux engines available:                 poll select
```

The `mpitest.py` script is as follows:
```python
from mpi4py.futures import MPIPoolExecutor

import time
import platform
import os
import sys

def task(_):
    start_time = time.time()
    result = fibonacci(start_time + 5)
    return result, start_time, f"{platform.node()} ({os.getpid()})"

def fibonacci(max_time):
    a = 0
    b = 1
    n = 1

    while time.time() <= max_time:
        a, b = b, a + b
        n += 1

    return n

if __name__ == "__main__":
    processes = int(sys.argv[-1])
    with open('output.txt', 'w') as f:
        initial_time = time.time()
        f.write(f"Starting at {initial_time}\n\n")
        f.flush()

        test = MPIPoolExecutor(max_workers=processes)
        results = test.map(task, range(processes))
        for n, (result, start_time, node) in enumerate(results):
            f.write(f"Task {n} started on {node} after {start_time-initial_time} seconds. Result is: {result}.\n")
            f.flush()

        f.write(f"\nCompleted after {time.time() - initial_time} seconds.\n")
```

The `nompi.py` script is the same, except that `MPIPoolExecutor` is replaced with `ProcessPoolExecutor`.

The job script without MPI is
```bash
#!/bin/bash
#SBATCH --partition=XXX
#SBATCH --account=XXX

#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=51
#SBATCH --mem-per-cpu=1G

#SBATCH --time=0-00:10

source ~/.bashrc

module purge
module load mpi/mpich-x86_64
conda activate test-env-mpich

python nompi.py 50
```

The job script with MPI is
```bash
#!/bin/bash
#SBATCH --partition=XXX
#SBATCH --account=XXX

#SBATCH --nodes=1
#SBATCH --ntasks=51
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=1G

#SBATCH --time=0-00:10

source ~/.bashrc

module purge
module load mpi/mpich-x86_64
conda activate test-env-mpich

mpirun -np $SLURM_NTASKS python -m mpi4py.futures mpitest.py 50
```

**Comment by dalcinl** [comment, marked as answer]
If could be that the MPICH package coming from conda-forge simply do not interact well with your SLURM. I do not remember whether MPICH has to be specially configured to support SLURM. If that's the case, then the conda-forge package is definitely not prepared to work under SLURM.

Could you `export HYDRA_TOPO_DEBUG=1` as explained [here](https://github.com/pmodels/mpich/blob/main/doc/wiki/how_to/Using_the_Hydra_Process_Manager.md) ? That would print the MPI process binding.
Another thing you could try is to NOT use `#SBATCH -cpus-per-task=1`, and instead use `mpirun -bind-to core` or `mpirun -bind-to hwthread`.

**Reply by pmenczel** [reply]
With the original script, `HYDRA_TOPO_DEBUG` was not doing anything. After changing the cpus-per-task option to `-bind-to core` or `-bind-to hwthread`, it does print the binding, and that slowdown is gone as well. Thank you very much for helping, even though the problem was not mpi4py-related at all! (And of course, thanks to everyone for your work here.)

One more weird thing, not sure if I should worry about this: when using only one node, it seems that some of the processes do not get a binding. The output looks like this:
```
process 0 binding: 100000000000000000000000000000000000000000000000000
process 1 binding: 010000000000000000000000000000000000000000000000000
...
process 25 binding: 000000000000000000000000010000000000000000000000000
process 26 binding: 000000000000000000000000000000000000000000000000000
...
process 50 binding: 000000000000000000000000000000000000000000000000000
```
In cases where I use more than one node, that doesn't seem to happen (there are no lines with all zeroes).

**Reply by dalcinl** [reply]
Weird indeed. I've not seen such output before. Are you running in some sort for container/virtual machine? It is surprising that your binding row has exactly 51 entries, which would be a weird figure for a physical CPU socket. To be honest, I have never printed before the process bindings running under SLURM, so I'm not sure if that may have some effect on it. It could very well be some minor bug on MPICH or the hwloc library. Maybe report it to https://github.com/pmodels/mpich/issues?

**Reply by pmenczel** [reply]
I do see 50 worker processes starting to run at the same time, and all on different cores, judging from the speed of the calculations. The number of entries per row is indeed always equal to the number of processes. If I use fewer processes, there are even cases where all entries in all rows are zero. (However, it still appears to be working just fine.)

It is an actual supercomputer with no VM and no container that I would be aware of. Every node should have 2 sockets with 56 cores each. I am just a user, so my access to configuration and information is limited. I'm not sure if I will be able to create a useful issue there, but I will give it a try.
